core/src/main/java/com/linecorp/armeria/server/docs/TypeSignature.java

```diff
@@ -0,0 +1,272 @@
+/*
+ *  Copyright 2017 LINE Corporation
+ *
+ *  LINE Corporation licenses this file to you under the Apache License,
+ *  version 2.0 (the "License"); you may not use this file except in compliance
+ *  with the License. You may obtain a copy of the License at:
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *  License for the specific language governing permissions and limitations
+ *  under the License.
+ */
+package com.linecorp.armeria.server.docs;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableList;
+
+/**
+ * Type signature of a method parameter, a method return value or a struct/exception field.
+ * A type signature can be represented as a string in one of the following forms:
+ * <ul>
+ *   <li>Base types: {@code "{type name}"}<ul>
+ *     <li>e.g. {@code "i64"}, {@code "double"} and {@code "string"}</li>
+ *   </ul></li>
+ *   <li>List or set: {@code "{type name}<{element type signature}>"}<ul>
+ *     <li>e.g. {@code "list<i32>"} and {@code "list<set<string>>"}</li>
+ *   </ul></li>
+ *   <li>Map: {@code "map<{key type signature}, {value type signature}>"}<ul>
+ *     <li>e.g. {@code "map<string, com.example.FooStruct>"}</li>
+ *   </ul></li>
+ *   <li>Named types (enums, structs and exceptions): {@code "{fully qualified type name}"}<ul>
+ *     <li>e.g. {@code "com.example.FooStruct"}</li>
+ *   </ul></li>
+ *   <li>Unresolved types: {@code "?{type name}"}<ul>
+ *     <li>e.g. {@code "?BarStruct"} and {@code "?com.example.BarStruct"}</li>
+ *   </ul></li>
+ * </ul>
+ */
+@JsonSerialize(using = TypeSignatureJsonSerializer.class)
+public final class TypeSignature {
```

Comment:
This PR has everything I've ever wanted, including ones I didn't ask for yet! :D